terracotta/drivers/driver.py

```diff
@@ -0,0 +1,177 @@
+        *args: Any, **kwargs: Any
+    ) -> T:
+        if requires_all_keys and (keys is None or len(keys) != len(self.key_names)):
+            raise exceptions.InvalidKeyError(
+                f'Got wrong number of keys (available keys: {self.key_names})'
+            )
+
+        if isinstance(keys, Mapping):
+            keys = dict(keys.items())
+        elif isinstance(keys, Sequence):
+            keys = dict(zip(self.key_names, keys))
+        elif keys is None:
+            keys = {}
+        else:
+            raise exceptions.InvalidKeyError(
+                'Encountered unknown key type, expected Mapping or Sequence'
+            )
+
+        unknown_keys = set(keys) - set(self.key_names)
+        if unknown_keys:
+            raise exceptions.InvalidKeyError(
+                f'Encountered unrecognized keys {unknown_keys} (available keys: {self.key_names})'
+            )
+
+        # Apparently mypy thinks fun might still be None, hence the ignore:
+        return fun(self, keys, *args, **kwargs)  # type: ignore
+    return inner
+
+
+class TerracottaDriver:
+
+    def __init__(self, metastore: MetaStore, rasterstore: RasterStore) -> None:
+        self.metastore = metastore
+        self.rasterstore = rasterstore
+
+        settings = terracotta.get_settings()
+        self.LAZY_LOADING_MAX_SHAPE: Tuple[int, int] = settings.LAZY_LOADING_MAX_SHAPE
+
+    @property
+    def db_version(self) -> str:
+        return self.metastore.db_version
+
+    @property
+    def key_names(self) -> Tuple[str, ...]:
+        return self.metastore.key_names
+
+    def create(self, keys: Sequence[str], *,
+               key_descriptions: Mapping[str, str] = None) -> None:
+        self.metastore.create(keys=keys, key_descriptions=key_descriptions)
+
+    def connect(self, verify: bool = True) -> contextlib.AbstractContextManager:
+        return self.metastore.connect(verify=verify)
+
+    @requires_connection
+    def get_keys(self) -> OrderedDict:
+        return self.metastore.get_keys()
+
+    @requires_connection
+    @standardize_keys
+    def get_datasets(self, keys: MultiValueKeysType = None,
+                     page: int = 0, limit: int = None) -> Dict[Tuple[str, ...], Any]:
+        return self.metastore.get_datasets(
+            where=keys,
+            page=page,
+            limit=limit
+        )
+
+    @requires_connection
+    @standardize_keys(requires_all_keys=True)
+    def get_metadata(self, keys: ExtendedKeysType) -> Dict[str, Any]:
+        keys = cast(KeysType, keys)
+
+        metadata = self.metastore.get_metadata(keys)
+
+        if metadata is None:
+            # metadata is not computed yet, trigger lazy loading
+            handle = only_element(self.get_datasets(keys).values())
+            metadata = self.compute_metadata(handle, max_shape=self.LAZY_LOADING_MAX_SHAPE)
+            self.insert(keys, handle, metadata=metadata)
+
+            # this is necessary to make the lazy loading tests pass...
```

Comment:
Probably because of small deviations in floating points... Alternatively because one of them (can't remember which) returned a list of tuples, whereas the other returned i list of lists. Not quite sure why, but that fixed a failing test.
But I'd certainly prefer to be without it -- it seems weird

Reply:
I think it would pay off to find out which one it is. If it's float precision then we should absolutely keep this to keep things consistent. If it's different data structures we should just change the test to allow for that, since JSON responses don't distinguish between list and tuple.

Reply:
Let's just adjust the comment then :)